File: BaekjoonOnlineJudge/1647/1647.py
import sys
input = sys.stdin.readline


def find(target):
    if disjointSet[target] == target:
        return target
    
    disjointSet[target] = find(disjointSet[target])
    return disjointSet[target]


def union(sv, dv):
    findSV = find(sv)
    findDV = find(dv)

    if findSV == findDV:
        return False

    if findSV < findDV:
        disjointSet[findDV] = findSV
    else:
        disjointSet[findSV] = findDV

    return True


def kruskal():
    global ans

    count = 0
    for w, sv, dv in roads:
        if union(sv, dv):
            count += 1

            if count < N - 1:
                ans += w


N, M = map(int, input().split())

ans = 0
disjointSet = [x for x in range(N + 1)]

roads = []
for _ in range(M):
    A, B, C = map(int, input().split())

    roads.append((C, A, B))
    roads.append((C, B, A))

roads.sort()

kruskal()

print(ans)

File: BaekjoonOnlineJudge/1647/1647.java
import java.io.*;
import java.util.*;

class Edge {

    int weight;
    int from;
    int to;

    public Edge(int weight, int from, int to) {
        this.weight = weight;
        this.from = from;
        this.to = to;
    }
}

public class Main {

    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    private static StringTokenizer st;

    private static int n, m;
    private static int[] disjointSet;
    private static PriorityQueue<Edge> edgePq = new PriorityQueue<>(
        (e1, e2) -> e1.weight - e2.weight);
    private static int answer = 0;

    private static int find(int target) {
        if (disjointSet[target] == target) {
            return target;
        }

        return disjointSet[target] = find(disjointSet[target]);
    }

    private static boolean union(int e1, int e2) {
        int findE1 = find(e1);
        int findE2 = find(e2);

        if (findE1 == findE2) {
            return false;
        }

        if (findE1 < findE2) {
            disjointSet[findE2] = findE1;
        } else {
            disjointSet[findE1] = findE2;
        }

        return true;
    }

    private static void kruskal() {
        int maxWeight = 0;

        while (!edgePq.isEmpty()) {
            Edge e = edgePq.poll();

            if (union(e.from, e.to)) {
                if (maxWeight < e.weight) {
                    maxWeight = e.weight;
                }

                answer += e.weight;
            }
        }

        answer -= maxWeight;
    }

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        disjointSet = new int[n + 1];

        for (int i = 1; i < n + 1; i++) {
            disjointSet[i] = i;
        }

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());

            int from = Integer.parseInt(st.nextToken());
            int to = Integer.parseInt(st.nextToken());
            int weight = Integer.parseInt(st.nextToken());

            edgePq.add(new Edge(weight, from, to));
        }

        kruskal();

        bw.write(String.valueOf(answer));

        bw.flush();
        bw.close();
        br.close();
    }
}